File: test/channelTest.cpp
#include "Channel.h"
#include "GMSKTranscoder.hpp"
#include "catch2/catch_all.hpp"
#include <fstream>
#include <iostream>
#include <string>

constexpr uint16_t PacketLength = 100;
constexpr uint16_t NumPackets = 60;
constexpr uint8_t MaxSNR = 20;
etl::bitset<NumPackets *PacketLength> gmskInput =
    etl::bitset<NumPackets * PacketLength>();

const uint8_t samplesPerSymbol = 10;
double gmskIout[samplesPerSymbol * PacketLength * NumPackets] = {0};
double gmskQout[samplesPerSymbol * PacketLength * NumPackets] = {0};



Channel channel;

TEST_CASE("TEST EFFECTS ON GMSK SIGNAL") {
    double gmskIoutTemp[samplesPerSymbol * PacketLength * NumPackets];
    double gmskQoutTemp[samplesPerSymbol * PacketLength * NumPackets];


    int sampleFrequency = 48000;
    int symbolRate = sampleFrequency / samplesPerSymbol;
    GMSKTranscoder<samplesPerSymbol, NumPackets * PacketLength> gmskMod(
        sampleFrequency, symbolRate, false);

    // Add NUM_PACKETS random packets to the input signal
    std::srand(std::time(nullptr));
    for (int i = 0; i < NumPackets; i++) {
        for (int j = 0; j < PacketLength; j++) {
            gmskInput.set(PacketLength * i + j, std::rand() % 2);
        }
    }

    gmskMod.modulate(gmskInput, gmskIout, gmskQout);

    for(int j = 0; j < samplesPerSymbol * PacketLength * NumPackets; j++){
        gmskIoutTemp[j] = gmskIout[j];
        gmskQoutTemp[j] = gmskQout[j];
    }

    // create a file with i/q data
    std::fstream channelFile;
    channelFile.open("../test/iofiles/gmskSignal.txt", std::ios::out);
    if (channelFile.is_open()) {
        for (int i = 0; i < samplesPerSymbol * PacketLength * NumPackets;
             i++) {
            channelFile << gmskIout[i] << " " << gmskQout[i] << " ";
        }
    }

    //AGWN TO GMSK SIGNAL
    double snrDB = 10;
    channel.addWGN(gmskIout, gmskQout, snrDB, symbolRate, samplesPerSymbol,
                   NumPackets, PacketLength);

    // create a file with i/q data
    std::fstream addWGNFile;
    addWGNFile.open("../test/iofiles/wgnSignal.txt", std::ios::out);
    if (addWGNFile.is_open()) {
        for (int i = 0; i < samplesPerSymbol * PacketLength * NumPackets;
             i++) {
            addWGNFile << gmskIout[i] << " " << gmskQout[i] << " ";
        }
    }

    for(int j = 0; j < samplesPerSymbol * PacketLength * NumPackets; j++){
        gmskIout[j] = gmskIoutTemp[j];
        gmskQout[j] = gmskQoutTemp[j];
    }


    // PHASE SHIFT GMSK SIGNAL
    channel.shiftPhase(gmskIout, gmskQout, samplesPerSymbol, NumPackets,
                       PacketLength, M_PI);

    std::fstream phaseShiftFile;
    phaseShiftFile.open("../test/iofiles/phaseShiftedSignal.txt",
                        std::ios::out);
    if (phaseShiftFile.is_open()) {
        for (int i = 0; i < samplesPerSymbol * PacketLength * NumPackets;
             i++) {
            phaseShiftFile << gmskIout[i] << " " << gmskQout[i] << " ";
        }
    }

    //FREQUENCY SHIFT GMSK SIGNAL
    double freqShift = 1000;
    channel.shiftFrequency(gmskIout, gmskQout, samplesPerSymbol, NumPackets,
                           PacketLength, freqShift, 2000);

    // create a file with i/q data
    std::fstream freqShiftFile;
    freqShiftFile.open("../test/iofiles/freqShiftedSignal.txt",
                       std::ios::out);
    if (freqShiftFile.is_open()) {
        for (int i = 0; i < samplesPerSymbol * PacketLength * NumPackets;
             i++) {
            freqShiftFile << gmskIout[i] << " " << gmskQout[i] << " ";
        }
    }
}

File: test/channelTest.py
import matplotlib.pyplot as plt


# TEST CASE: CREATE GMSK SIGNAL
with open('iofiles/gmskSignal.txt') as f:
    signal = f.read().split()

i = []
q = []
for j in range(len(signal)):
    if j % 2 == 0:
        i.append(float(signal[j]))
    else:
        q.append(float(signal[j]))

plt.plot(i[0:160], 'r-')
plt.plot(q[0:160], 'b-')

plt.title("GMSK SIGNAL")
plt.show()

# TEST CASE: ADD WGN TO THE GMSK SIGNAL

with open('iofiles/wgnSignal.txt') as f:
    wgnSignal = f.read().split()

iWGN = []
qWGN = []
for j in range(len(wgnSignal)):
    if j % 2 == 0:
        iWGN.append(float(wgnSignal[j]))
    else:
        qWGN.append(float(wgnSignal[j]))

plt.plot(iWGN[0:160], 'y-', qWGN[0:160], 'g-')
plt.plot(i[0:160], 'r-')
plt.plot(q[0:160], 'b-')
plt.title("WGN added to GMSK Signal")

plt.show()

# TEST CASE: PHASE SHIFT THE GMSK SIGNAL

with open('iofiles/phaseShiftedSignal.txt') as f:
    phaseShiftedSignal = f.read().split()

iPhaseShifted = []
qPhaseShifted = []
for j in range(len(phaseShiftedSignal)):
    if j % 2 == 0:
        iPhaseShifted.append(float(phaseShiftedSignal[j]))
    else:
        qPhaseShifted.append(float(phaseShiftedSignal[j]))

plt.plot(iPhaseShifted[0:160], 'g-')
plt.plot(qPhaseShifted[0:160], 'y-')
plt.plot(i[0:160], 'r-')
plt.plot(q[0:160], 'b-')

plt.title("Phase shifted GMSK signal")

plt.show()

# TEST CASE: FREQUENCY SHIFT THE GMSK SIGNAL

with open('iofiles/freqShiftedSignal.txt') as f:
    freqShiftedSignal = f.read().split()

iFreqShifted = []
qFreqShifted = []
for j in range(len(freqShiftedSignal)):
    if j % 2 == 0:
        iFreqShifted.append(float(freqShiftedSignal[j]))

    else:
        qFreqShifted.append(float(freqShiftedSignal[j]))

plt.plot(iFreqShifted[0:160], 'y-')
plt.plot(qFreqShifted[0:160], 'g-')
plt.plot(i[0:160], 'r-')
plt.plot(q[0:160], 'b-')

plt.title("Frequency shifted GMSK signal")

plt.show()
